horovod/spark/keras/datamodule.py

```diff
@@ -0,0 +1,173 @@
+# Copyright (C) 2022, NVIDIA CORPORATION. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+import contextlib
+import horovod.tensorflow.keras as hvd
+
+from horovod.spark.common import constants
+from horovod.spark.common.datamodule import DataModule
+
+
+class PetastormDataModule(DataModule):
+    """Default Petastorm-based DataModule for KerasEstimator."""
+    def __init__(self, reader_pool_type: str="process",
+                       train_reader_worker_count: int=2,
+                       val_reader_worker_count: int=2,
+                       make_dataset=None,
+                       shuffle=False,
+                       random_seed=0,
+                       **kwargs):
+        from petastorm import TransformSpec, make_reader, make_batch_reader
+
+        super().__init__(**kwargs)
+        self.reader_pool_type = reader_pool_type
+        self.train_reader_worker_count = train_reader_worker_count
+        self.val_reader_worker_count = val_reader_worker_count
+        self.make_dataset = make_dataset
+        self.shuffle = shuffle
+        self.random_seed = random_seed
+
+        # In general, make_batch_reader is faster than make_reader for reading the dataset.
+        # However, we found out that make_reader performs data transformations much faster than
+        # make_batch_reader with parallel worker processes. Therefore, the default reader
+        # we choose is make_batch_reader unless there are data transformations.
+        self.transform_spec = TransformSpec(self.transform_fn) if self.transform_fn else None
+        self.reader_factory_kwargs = dict()
+        if self.transform_spec:
+            self.reader_factory = make_reader
+            self.reader_factory_kwargs['pyarrow_serialize'] = True
+            self.is_batch_reader = False
+        else:
+            self.reader_factory = make_batch_reader
+            self.is_batch_reader = True
+
+    def __enter__(self):
+        super().__enter__()
+        self.train_reader = self.reader_factory(self.train_dir,
+                                num_epochs=self.num_train_epochs,
```

Comment:
`num_epochs=1`
To be consistent with https://github.com/horovod/horovod/blob/master/horovod/spark/keras/remote.py#L253

Petastorm will repeat internally at https://github.com/horovod/horovod/blob/master/horovod/spark/keras/util.py#L82